docs/proposals/scheduling/20220629-schedule-device-in-card-level.md

```diff
@@ -0,0 +1,143 @@
+need 1c.2g.20gb, we can't expect the pods requirements in advance, so there is a demand for scheduler to make the 
+best placement for ci\gi-create and schedule pod in it.
+
+#### Story Oversold
+in some cases, we find although some cards free is zero in scheduler account, but we observed that the physical
+utilization of gpu-mem\gpu-core still has buffer, so we may schedule some low-priority jobs to use this buffer, and
+when buffer disappear, we should kill these jobs to avoid affecting normal jobs. because the utilization is different
+bewteen cards, so scheduler must schedule based card level.
+     
+
+## Implementation Details
+
+#### Scheduler
+1. we need a new data structure to describe each device card resources and healthy status on one node
+2. we need new plugins to execute filter\reserve\bind stage
+3. support user to apply any ratio of one card, like 50%/25%
+4. automatically recognize different kind devices. when a new device added, we don't need modify any code
+5. schedule-result will be written on pod annotations.
+
+#### device reporter
+we need create a new device reporter on node to update a new device crd to report each device card resource and health status
+
+```go
+type DeviceType string
+
+const (
+	GPU  DeviceType = "gpu-device"
+	NPU  DeviceType = "npu-device"
+	FPGA DeviceType = "fpga-device"
+	RDMA DeviceType = "rdma-device"
+	POV  DeviceType = "pov-device"
+)
+
+type DeviceList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items  []Device `json:"items"`
+}
+
+type Device struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	NodeName     string        `json:"nodeName"`
+	DeviceType   DeviceType    `json:"deviceType"`
+	Spec         DeviceSpec    `json:"spec,omitempty"`
+}
+
+type DeviceSpec struct {
+	DeviceInfos     []DeviceInfo    `json:"deviceInfos"`
+    deviceTopology  DeviceTopology  `json:"deviceTopology"`
+}
+
+type DeviceInfo struct {
+	ID        string                       `json:"id"`       // Device UUID
+	Resources map[string]resource.Quantity `json:"resource"` // map of sub-resources
+	Health    bool                         `json:"health"`
+	Minor     int32                        `json:"minor"` // index starting from 0
+}
+
+type DeviceTopology struct {
+	// 2D array of p2p linkMeta, indexed by device minor
+	Links [][]GpuLink `json:"links,omitempty"`
+}
+type DeviceLink struct {
+	Type      string      `json:"LinkType,omitempty"`
+	Bandwidth int64       `json:"bandwidth,omitempty"`
+}
+```
+
+### RuntimeProxy
+our target is to work well with origin k8s kubelet and k8s device plugins, so:
+1. we still allow kubelet and device plugin to allocate concrete device card.
+2. in runtimeproxy, we will use pod's devive-allocation-result in annotation to replace the step1's result of container's args and envs
```

Comment:
If doing this way, l think in runtime layer you need to know the exact env name like `NVIDIA_VISIBLE_DEVICES `, then you can overwrite the allocation, but this is different for different devices, how should you know that?

Reply:
different devices have relatively fixed pattern and less likely to change due to compatibility reason

Reply:
Thanks. But in addition to these environment variables, some device plugins also set the mount and devices of the container, modifying them at a lower level may need to do some dirty work 🤔

Reply:
We don't have to consider supporting the existing DevicePlugin. As we discussed before, we can report device capacity through the koordlet.

Reply:
yeah l know you can collect device info through koordlet and make allocation decision in scheduler, but you have to know the exact ways to set correct device envs, mounts and devices in runtime layer, just as what device plugin does

Reply:
In order to add devices without changing the code, we may need another device plugins framework.

Reply:
> In order to add devices without changing the code, we may need another device plugins framework.

+1